tests/test_run_service.py

```diff
@@ -0,0 +1,475 @@
+                logger.info(f"Waiting {retry_delay} seconds for containers to start...")
+                time.sleep(retry_delay)
+                continue
+            
+            # Verify all running containers are actually running
+            all_running = all(c.status == "running" for c in running_containers)
+            if all_running:
+                logger.info("All trader containers are running")
+                return True
+            
+            if attempt == max_retries - 1:
+                return False
+                
+            logger.info(f"Some containers not running, waiting {retry_delay} seconds...")
+            time.sleep(retry_delay)
+            
+        except Exception as e:
+            logger.error(f"Error checking Docker status: {str(e)}")
+            if attempt == max_retries - 1:
+                return False
+            logger.info(f"Waiting {retry_delay} seconds before retry...")
+            time.sleep(retry_delay)
+    
+    return False
+
+def check_service_health(logger: logging.Logger) -> tuple[bool, dict]:
+    """Enhanced service health check with metrics, monitoring for 2 minutes."""
+    metrics = {
+        'response_time': None,
+        'status_code': None,
+        'error': None,
+        'check_count': 24,  # Total number of checks (2 minutes / 5 seconds)
+        'successful_checks': 0
+    }
+    
+    start_monitoring = time.time()
+    while time.time() - start_monitoring < 120:  # Run for 2 minutes
+        try:
+            start_time = time.time()
+            response = requests.get(HEALTH_CHECK_URL, timeout=10)
+            metrics['response_time'] = time.time() - start_time
+            metrics['status_code'] = response.status_code
+            
+            if response.status_code == 200:
+                metrics['successful_checks'] += 1
+                logger.info(f"Health check passed (response time: {metrics['response_time']:.2f}s)")
+            else:
+                logger.error(f"Health check failed - Status: {response.status_code}")
+                return False, metrics
+                
+        except requests.exceptions.Timeout:
+            metrics['error'] = 'timeout'
+            logger.error("Health check timeout")
+            return False, metrics
+        except requests.exceptions.ConnectionError as e:
+            metrics['error'] = 'connection_error'
+            logger.error(f"Connection error: {str(e)}")
+            return False, metrics
+        except Exception as e:
+            metrics['error'] = str(e)
+            logger.error(f"Unexpected error in health check: {str(e)}")
+            return False, metrics
+            
+        # Wait for remaining time in 5-second interval
+        elapsed = time.time() - start_time
+        if elapsed < 5:
+            time.sleep(5 - elapsed)
+    
+    # Return True only if all checks were successful
+    is_healthy = metrics['successful_checks'] == metrics['check_count']
+    return is_healthy, metrics
```

Comment:
we can directly return True here right? because we're already returning False in all other paths.
I'm concerned because it's possible that it only iterates 23 times because of some lag in between 120 seconds.

Reply:
**For health check metrics:
"I've updated the health check function to use a success rate approach instead of exact check count. Now it:**


Tracks actual checks performed
Calculates a success rate
Returns True if success rate ≥ 90%
Logs detailed metrics including success rate
This handles the case where we might get fewer iterations due to lag."

Reply:
thats very narrow edge case, bt thats safe hands, lemme update it that way then

Reply:
updated